sirepo/feature_config.py

```diff
@@ -120,10 +120,9 @@ def b(msg, dev=False):
             'Names of root packages that should be checked for codes and resources. Order is important, the first package with a matching code/resource will be used. sirepo added automatically.',
         ),
         proprietary_sim_types=(set(), set, 'codes that require authorization'),
-        raydata=dict(
-            # POSIT: Included in sirepo.job_driver.docker.cfg.aux_volumes
-            data_dir=(None, pkio.py_path, 'abspath of dir to store raydata analysis output')
-        ),
+        # SECURITY: This path should not be exposed to the user so do
```

Comment:
I think how we configure paths shouldn't be exposed to users. It gives them information about how our system is organized that they don't need to know and that could help them attack us. OTOH maybe so much is exposed through rsconf that it doesn't really matter. What do you think?

Reply:
RSConf config or it's output (e.g. `env`) is not visible to the user so I think that's a non-issue.

In this particularly case, I guess the config hits the javascript(?) so you're concerned about that?

While I appreciate hiding config (in general), I think we are in the gray area of hiding too much at the expense of added complexity (I like the modularized cfg in sub-dict `raydata`). One of the things we don't do well is protect error messages or output from code. If the notebook outputs something about the images (eg their full path), then the information will be exposed. Since we have moved to (almost :) full execution in agents, I've been less concerned about path security. Once arbitrary code can be executed (uploaded notebooks), the entire container's file system is visible, and therefore modifiable unless specifically mounted read-only (which data_dir can't be). 

In this particular application, we have to assume the users are trusted. We do this anyway, since there is no concept of user in Bluesky. If there was, we would only mount that user's files in the container, and we'd buffer all accesses through controlling credentials to Bluesky in the server. If this happens in a future version, the root of the directory can be known, because it would not be mounted in the container, just the user's directory.

Reply:
> RSConf config or it's output (e.g. env) is not visible to the user so I think that's a non-issue.

True. My thinking was that rsconf still exposes some info. How much is too much I have no clue (other than the obvious like a password). For example, rsconf exposes that proprietary codes live in a `proprietary_codes` dir that are at one level down under the run_dir. I don't know that this info is all that useful but it may be enough for an attacker to say "well I just need to be able to reach *this* level in the filesystem and I can get at useful info".

> In this particularly case, I guess the config hits the javascript(?) so you're concerned about that?

Yes. If we do the way of cfg in a sub-dict then the user can type `SIREPO.APP_SCHEMA.feature_config` in the console and see the path.

> If the notebook outputs something about the images (eg their full path), then the information will be exposed

Their full path inside of the docker container will be exposed. With the raydata.data_dir the path on the host machine will be exposed.

Reply:
> For example, rsconf exposes that proprietary codes live in a proprietary_codes dir that are at one level down under the run_dir.

The point you are making is subtle. If someone wants to hack Sirepo, you are saying they can extract a lot of information from Sirepo, Pykern, and RSConf, because we write open source software, and we don't try to hide or configure everything on the assumption that the host system can be secured reasonably. This is, perhaps, over-reliance on Docker. Knowing a host path doesn't do much, if you can't inject code on the host. That's the premise here.

> "well I just need to be able to reach this level in the filesystem and I can get at useful info"

The question is what is useful. RSConf exposes `env` in the source code of RSConf. If we use Vault or something else, it's really just a bit of an abstraction that defers access. I don't think it fixes anything when, say, someone can inject code. Rather, knowing app uses Vault gives a lot of information. Using microservices can further obfuscate things, but I don't think it ensures much if someone can penetrate the system.

Originally I wanted to hide all paths, but given the nature of being a scientific gateway for complex codes, it's become harder. Scientists need all the info they can get to debug. 

> Their full path inside of the docker container will be exposed. With the raydata.data_dir the path on the host machine will be exposed.

I think that's generally true, because we almost always mount the same name inside the agent (except NERSC where we don't control things). This helps debugging. There's a difference between exposing a path and having access to the path, even if it does give a clue to how to hack the system. 

In this particular case, this is a behind-the-firewall implementation for a particular class of database that we can't protect. I don't think knowing paths is going to change the level of security of the whole system or even in leaving a "bad example" for others to follow.

That said, this is pretty abstract so make a choice, and I'll approve. :)

Reply:
Thanks for all the details:) 

> Originally I wanted to hide all paths, but given the nature of being a scientific gateway for complex codes, it's become harder. Scientists need all the info they can get to debug.

I guess I was still working under the "hide all paths" model. If you're cool exposing the path then I'm happy to work that way.

> In this particular case, this is a behind-the-firewall implementation for a particular class of database that we can't protect.
That makes sense. In cases for other codes should I be concerned about exposing paths? What about paths that contain the user's sirepo uid?

Reply:
bf96aecc7

Reply:
> If you're cool exposing the path then I'm happy to work that way.

I think each time we do this, we need to look at it. Security is a process, not a rule.

> In cases for other codes should I be concerned about exposing paths?

I don't think so in general, because run_dir gets exposed. However, see note above. :)

> What about paths that contain the user's sirepo uid?

That's an interesting one. run_dir contains uid and sid. I don't know what we can do about this than stripping all output. That's hard to do, because it comes in many forms. I think in general we should be careful where we can, but until we fix the general output filtering problem, there will be leakage.

Reply:
"process not a rule" I like it.

What you said sounds good.